pkg/kubectl/cmd/portforward.go

```diff
@@ -118,20 +125,43 @@ func (o *PortForwardOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, arg
 	var err error
 	o.PodName = cmdutil.GetFlagString(cmd, "pod")
 	if len(o.PodName) == 0 && len(args) == 0 {
-		return cmdutil.UsageErrorf(cmd, "POD is required for port-forward")
+		return cmdutil.UsageErrorf(cmd, "POD or RESOURCE/NAME is required for port-forward")
+	}
+
+	o.Namespace, _, err = f.DefaultNamespace()
+	if err != nil {
+		return err
 	}
 
+	builder := f.NewBuilder().
+		Internal().
+		ContinueOnError().
+		NamespaceParam(o.Namespace).DefaultNamespace()
+
 	if len(o.PodName) != 0 {
 		printDeprecationWarning("port-forward POD", "-p POD")
 		o.Ports = args
 	} else {
-		o.PodName = args[0]
-		o.Ports = args[1:]
-	}
+		getPodTimeout, err := cmdutil.GetPodRunningTimeoutFlag(cmd)
+		if err != nil {
+			return cmdutil.UsageErrorf(cmd, err.Error())
+		}
 
-	o.Namespace, _, err = f.DefaultNamespace()
-	if err != nil {
-		return err
+		resourceName := args[0]
+		builder.ResourceNames("pods", resourceName)
+
+		obj, err := builder.Do().Object()
+		if err != nil {
+			return err
+		}
+
+		forwardablePod, err := f.AttachablePodForObject(obj, getPodTimeout)
+		if err != nil {
+			return err
+		}
+
+		o.PodName = forwardablePod.Name
+		o.Ports = args[1:]
```

Comment:
If we port forward a service, wouldn't the specified port be the service port, and would we need to figure out what port on the resolved pod that mapped to?

Reply:
It is likely the intended port would be the service port, but then the user still need to specify what local port to bind to so it is not clear to me what it can help here.  Also it is possible that a service has multiple service ports, and the user may not want to bind all of them.   I feel it would end up complicate the UI and not simplify it.

The same mechanism also supports other resource types such as statefulset, deployment, or job, so it is probably cleaner to make less assumptions about the ports.

Reply:
> then the user still need to specify what local port to bind to so it is not clear to me what it can help here

The user already has the ability to specify local/remote mappings. I'm saying the remote port should resolve the service port to determine what pod port to forward to, not assume the port number specified by the user is the port to use on the pod. 

> The same mechanism also supports other resource types such as statefulset, deployment, or job, so it is probably cleaner to make less assumptions about the ports.

None of those provide mechanisms for port indirections like services do

Reply:
I see, let's me see if I get it right this time.  I think you are suggesting that it should also resolve spec.ports.port -> spec.ports.targetPort if resource type is a service.

In that case, I think we overload the design of port-forward to a pod to become port-forward to a service, and I am not sure if it should be in the scope of this PR.

Reply:
> I think you are suggesting that it should also resolve spec.ports.port -> spec.ports.targetPort if resource type is a service.

Yes

> I am not sure if it should be in the scope of this PR.

Agree. It could be a follow up. I think it is useful enough to consider doing, and a natural fit for this command.

Reply:
I will enter an issue to track that.

Reply:
I think there already is one.